src/util/codec/table.rs

```diff
@@ -229,43 +229,116 @@ pub trait TableDecoder: DatumDecoder {
-    // hack: HashMap will still allocate memeory when capacity is 0, need to use new instead.
-    if cols.is_empty() {
-        return Ok(HashMap::new());
-    }
-    let mut res = HashMap::with_capacity(cols.len());
-    if data.is_empty() || data.len() == 1 && data[0] == datum::NIL_FLAG {
-        return Ok(res);
-    }
-    while !data.is_empty() && res.len() < cols.len() {
-        let id = try!(data.decode_datum()).i64();
-        let (val, rem) = try!(datum::split_datum(data, false));
-        if cols.contains(&id) {
-            res.insert(id, val);
+pub struct RowColMeta {
+    offset: usize,
+    length: usize,
+}
+
+pub struct RowColsDict {
+    // data of current row
+    value: Vec<u8>,
+    // cols contains meta of each column in the format of:
+    // (col_id1,(offset1,len1)),(col_id2,(offset2,len2),...)
+    cols: HashMap<i64, RowColMeta>,
+}
+
+impl RowColMeta {
+    pub fn new(offset: usize, length: usize) -> RowColMeta {
+        RowColMeta {
+            offset: offset,
+            length: length,
         }
-        data = rem;
     }
-    Ok(res)
 }
 
-pub type ValDict<'a> = HashMap<i64, &'a [u8]>;
+impl RowColsDict {
+    fn new(cols_num: usize, val: Vec<u8>) -> RowColsDict {
+        RowColsDict {
+            value: val,
+            cols: HashMap::with_capacity(cols_num),
+        }
+    }
 
-// `cut_idx_key` cuts encoded index key into colIDs to bytes slices map.
-pub fn cut_idx_key<'a>(mut key: &'a [u8], col_ids: &[i64]) -> Result<(ValDict<'a>, &'a [u8])> {
-    key = &key[PREFIX_LEN + ID_LEN..];
-    if col_ids.is_empty() {
-        return Ok((HashMap::new(), key));
+    fn set_value(&mut self, val: Vec<u8>) {
+        self.value = val;
     }
-    let mut values = HashMap::with_capacity(col_ids.len());
-    for &id in col_ids {
-        let (val, rem) = try!(datum::split_datum(key, false));
-        values.insert(id, val);
-        key = rem;
+
+    fn insert(&mut self, key: i64, offset: usize, length: usize) -> Option<RowColMeta> {
+        let value = RowColMeta::new(offset, length);
+        self.cols.insert(key, value)
+    }
+
+    pub fn len(&self) -> usize {
+        self.cols.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.cols.is_empty()
+    }
+
+    pub fn get(&self, key: i64) -> Option<&[u8]> {
+        if let Some(meta) = self.cols.get(&key) {
+            return Some(&self.value[meta.offset..(meta.offset + meta.length)]);
+        }
+        None
     }
-    Ok((values, key))
+}
+
+// `cut_row` cut encoded row into (col_id,offset,length)
+// and return interested columns' meta in RowColsDict
+pub fn cut_row(data: Vec<u8>, cols: &HashSet<i64>) -> Result<RowColsDict> {
+    if cols.is_empty() || data.is_empty() || data.len() == 1 && data[0] == datum::NIL_FLAG {
+        return Ok(RowColsDict::new(0, data));
+    }
+
+    // parse cols from data
+    let mut res = {
+        let mut res = RowColsDict::new(cols.len(), Vec::new());
```

Comment:
How about init `RowColsDict` with `data`, and add a `get_value() -> &[u8]` method?